script.js

```diff
@@ -1,4 +1,139 @@
+//let's define player modes
+var currentplayer = 1;
+// let's create a database to store the players' numbers. Don't forget to reset this to 0 whenever you restart the game.
+var playernumber = [];
+var player1number = playernumber[0];
+var player2number = playernumber[1];
+//let's define the game modes
+var rolldicemode = "Roll the dice";
+var chooseordermode = "Choose order of dice";
+var currentGameMode = rolldicemode;
+// let's create a database to record the dice numbers rolled.
+var diceNumbersRolled = [];
+var diceRoll1 = "";
+var diceRoll2 = "";
+// set output value as a global variable.
+var myOutputValue = "";
```

Comment:
I would strongly recommend you function scope `myOutputValue`, instead of leaving it as a global variable as you are doing now.

1. The `main` function will return `myOutputValue` into the HTML grey output box. This is the function which references `myOutputValue` the most, because it needs to change the output depending on the game state. 
2. Following Point 1, for every other function that is intended to return a string for `main`'s `myOutputValue`, `myOutputValue` needs not be processed by said functions as function parameters. Let's take a look at the definition of the `announceWinner` function on [lines 114 to 123](https://github.com/rocketacademy/basics-beat-that/blob/ceb9d40bec77da3db1d7eeefe6332919ace7570d/script.js#L114-L123), for instance. Let's assume my "perfect way of writing the function", which will be:
```
var announceWinner = function () {
  // I am deliberately naming this variable output, to differentiate
  // it from the myOutputValue in the main function. You can continue
  // naming it myOutputValue here if you wish, because it will still be
  // scoped to the announceWinner function instead of main if you use
  // 'var' keyword. However, I wanted to demonstrate that we can change
  // the myOutputValue string in the main function without passing it in as
  // a parameter to the 'annouceWinner' function, nor using myOutputValue
  // as a global variable.
  // The default value: Player 1 wins
  var output = "Player 1 wins! To play again, click submit.";
  if (playernumber[1] > playernumber[0]) {
    // Player 2 wins only if playernumber[1] > playernumber[0]
    output = "Player 2 wins! To play again, click submit.";
  }
  // The 3 assignment lines below assign values to
  // the 3 global variables: (1) currentGameMode,
  // (2) currentplayer, and (3) playernumber.
  // We can keep them as global variables for now,
  // since we want these values to persist across the
  // different game states.
  currentGameMode = rolldicemode;
  currentplayer = 1;
  playernumber = [];
  // This is the most important part of my argument. I am returning
  // the 'output' variable to be used in the 'main' function when we execute
  // the function, then assign it back to 'myOutputValue' in 'main' function.
  // Notice that by the time I can still return  'output' as a string, without having
  // to access 'myOutputValue' as a global variable, or as a function parameter
  // to the 'announceWinner' function?
  return output;
};
```

If you are still "processing" the above at this point, I hope [line 70](https://github.com/rocketacademy/basics-beat-that/blob/ceb9d40bec77da3db1d7eeefe6332919ace7570d/script.js#L70) clarifies why the above works, and why we do not need to pass in `myOutputValue` as a function parameter like `myOutputValue = announceWinner(myOutputValue)`, nor use `myOutputValue` as a global variable.

I will be pasting the contents of [line 70](https://github.com/rocketacademy/basics-beat-that/blob/ceb9d40bec77da3db1d7eeefe6332919ace7570d/script.js#L70) here, with my own comments explaining it.

```
...
...
...
else if ((currentGameMode = "Announce Winner")) {
  // You are executing the 'announceWinner' function,
  // and then taking its 'return' value (which is the 'return output;'
  // line in the example above) and assigning it to myOutputValue.
  // This means the string that is stored in 'output' by the time you
  // reach the 'return output;' line at the end of the 'announceWinner'
  // function will be stored into 'myOutputValue'. Now, do you not notice
  // this 'output' string is not dependent on / referencing any current 
  // value(s) stored in 'myOutputValue'? 'output' is an entire new string
  // in the 'announceWinner' function.
  myOutputValue = announceWinner();
}
...
...
...
```

Now if you apply this same strategy across the other game states (especially if you intend to modularise each game state into its own function), you will notice the same pattern. You can have those game state functions `return` strings to change the contents in `myOutputValue` (by storing said strings `return`ed as a result of executing those functions into `myOutputValue`).

Let's assume a hypothetical function for the `chooseordermode`game state. Your logic for the `choosegamemode` game state is [between line 33 and line 69](https://github.com/rocketacademy/basics-beat-that/blob/ceb9d40bec77da3db1d7eeefe6332919ace7570d/script.js#L33-L69). Now, I will refactor it this way:

```
else if (currentGameMode == chooseordermode) {
  // this time round, I will be passing in 'input' as
  // the parameter, because 'input' only exists in the
  // scope of the 'main' function. 'input' is a parameter
  // to the 'main' function. In the next example, I will
  // explain more why we need 'input' as the parameter.
  myOutputValue = chooseOrderMode(input);
}
```

And then, the chooseOrderMode function (definition) should look like this:

```
var chooseOrderMode = function (input) {
  // If player chooses 1st dice number as first numeral
  // CX: again, you can name it whatever you want if you do
  // not like 'output'. It can even be called 'myOutputValue',
  // but because we have the 'var' keyword here ('var' implies
  // you are creating a variable), JavaScript knows you are
  // scoping it to the 'chooseOrderMode' function, as opposed
  // to using an identically named variable 'myOutputValue' from
  // the global scope
  var output = "Your combined number is " +
    diceRoll1 +
    diceRoll2 +
    ".<br> <br> Click submit to continue.";
  var combinednumber = String(diceRoll1) + String(diceRoll2);

  // If player chooses 2nd dice number as first numeral
  // CX: this is a good example of why then, we need 'input'
  // as the parameter to the 'chooseOrderMode' function.
  // This is because we need access to the value of 'input' 
  // from 'main' around the time the execution of 'chooseOrderMode'
  // happens in the previous code example. We need the current
  // value of 'input' from 'main' to compare it directly with 'diceRoll2',
  // which you are doing in the if condition below.
  if (input == diceRoll2) {
    output = "Your combined number is " +
      diceRoll2 +
      diceRoll1 +
      ".<br> <br> Click submit to continue.";
    combinednumber = String(diceRoll2) + String(diceRoll1);
  }

  // let's push the combined number of the player into our database
  playernumber.push(combinednumber);
  console.log("player numbers array: " + playernumber);
  console.log("output: " + myOutputValue);
  console.log("combined number: " + combinednumber);
  // let's return the game mode to roll dice mode.
  currentGameMode = rolldicemode;
  // don't forget to reset the database of diceNumberRolled, because this will affect the value of the combined number which is made of up DiceRoll1 and DiceRoll2.
  diceNumbersRolled = [];

  // This is how we restart the round for player 2
  if (currentplayer == 2) {
    // change game mode to decide which is winner
    currentGameMode = "Announce Winner";
  } else {
    currentplayer = currentplayer + 1;
  }
};
```

Super verbose comment I know, but you can [review this lesson about global variables](https://basics.rocketacademy.co/8-managing-state-and-input-validation/8.1-program-lifecycle-and-state) if you are still not confident about the concept!